extras/module-node/duk_module_node.c

```diff
@@ -0,0 +1,248 @@
+/*
+ * Node.js-like require() and module loader
+ *
+ * https://nodejs.org/api/modules.html
+ */
+
+#include "duktape.h"
+#include "duk_module_node.h"
+
+static duk_int_t duk__eval_module_helper (duk_context *ctx);
+static void duk__push_module_object (duk_context *ctx, const char *id);
+
+static duk_resolve_module_function resolve_module_cb;
+static duk_load_module_function load_module_cb;
+
+static duk_bool_t duk__get_cached_module(duk_context *ctx, const char *id) {
+	duk_push_global_stash(ctx);
+	duk_get_prop_string(ctx, -1, "\xffModuleCache");
+	if (duk_get_prop_string(ctx, -1, id)) {
+		duk_remove(ctx, -2);
+		duk_remove(ctx, -2);
+		return 1;
+	}
+	else {
+		duk_pop_3(ctx);
+		return 0;
+	}
+}
+
+/*
+ * NOTE: duk__put_cached_module() leaves the module object on the
+ * stack as a convenience to the caller.
+ */
+static void duk__put_cached_module(duk_context *ctx) {
+	/* ... module */
+
```

Comment:
Trivia (very): in representing value stack shape I've usually used brackets to make the top and bottom explicit:

    [ foo bar quux]

and when the bottom is not known or isn't interesting:

    [ ... bar quux ]